File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int i = 0;
        int j = 0;
        int k = matrix.size()-1;
        int l = matrix[0].size()-1;
        int lengthlit = l + 1;
        int mid = 0;
        int midi = 0;
        int midj = 0;
        while(i * lengthlit + j <= k * lengthlit + l ){
            mid = ((i * lengthlit + j) + (k * lengthlit + l))/2;
            midi = mid / lengthlit;
            midj = mid % lengthlit;

            if(matrix[midi][midj] > target){
                mid--;
                k = mid / lengthlit;
                l = mid % lengthlit;
            }
            else if(matrix[midi][midj] < target){
                mid++;
                i = mid / lengthlit;
                j = mid % lengthlit;
            }
            else{
                return true;
            }

        }
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        i = 0
        j = 0
        k = len(matrix) - 1
        l = len(matrix[0]) - 1
        lengthlit = len(matrix[0])

        while (i* lengthlit) + j <= k * lengthlit + l:
            mid = (((lengthlit * i) + j) + ((lengthlit*k) + l))//2 
            mii = mid // lengthlit
            mij = mid % lengthlit

            if target < matrix[mii][mij]:
                mid -= 1
                k = mid // lengthlit
                l = mid % lengthlit
            elif target > matrix[mii][mij]:
                mid += 1
                i = mid // lengthlit
                j = mid % lengthlit
            else:
                return True
        return False
